src/NBitcoin/BlockStake.cs

```diff
@@ -183,24 +185,45 @@ public override bool IsProtocolTransaction()
         }
     }
 
-    public class ProvenHeaderConsensusFactory : PosConsensusFactory
-    {
-        public override BlockHeader CreateBlockHeader()
-        {
-            return base.CreateProvenBlockHeader();
-        }
-    }
-
     /// <summary>
     /// The consensus factory for creating POS protocol types.
     /// </summary>
     public class PosConsensusFactory : ConsensusFactory
     {
+        /// <summary>
+        /// A dictionary for types assignable from <see cref="ProvenBlockHeader"/>.
+        /// </summary>
+        private readonly ConcurrentDictionary<Type, bool> isAssignableFromProvenBlockHeader = new ConcurrentDictionary<Type, bool>();
+
+        /// <summary>
+        /// The <see cref="ProvenBlockHeader"/> type.
+        /// </summary>
+        private readonly TypeInfo provenBlockHeaderType = typeof(ProvenBlockHeader).GetTypeInfo();
```

Comment:
What I am not sure about this PR is whether this should go in the base type (then it will be picked by the `DefaultConsensusFactory`) which is more correct @fassadlr thoughts?